trie/stacktrie.go

```diff
@@ -26,28 +26,62 @@ import (
 )
 
 var (
-	ErrCommitDisabled = errors.New("no database for committing")
-	stPool            = sync.Pool{New: func() any { return new(stNode) }}
-	_                 = types.TrieHasher((*StackTrie)(nil))
+	stPool = sync.Pool{New: func() any { return new(stNode) }}
+	_      = types.TrieHasher((*StackTrie)(nil))
 )
 
-// NodeWriteFunc is used to provide all information of a dirty node for committing
-// so that callers can flush nodes into database with desired scheme.
-type NodeWriteFunc = func(path []byte, hash common.Hash, blob []byte)
+// StackTrieOptions contains the configured options for manipulating the stackTrie.
+type StackTrieOptions struct {
+	Writer  func(path []byte, hash common.Hash, blob []byte) // The function to commit the dirty nodes
+	Cleaner func(path []byte)                                // The function to clean up dangling nodes
+
+	SkipLeftBoundary  bool                                             // Flag whether the nodes on the left boundary are skipped for committing
+	SkipRightBoundary bool                                             // Flag whether the nodes on the right boundary are skipped for committing
+	OnBoundary        func(path []byte, hash common.Hash, blob []byte) // Callback invoked when the node is skipped committing
+}
+
+// NewStackTrieOptions initializes an empty options for stackTrie.
+func NewStackTrieOptions() *StackTrieOptions { return &StackTrieOptions{} }
+
+// WithWriter configures trie node writer within the options.
+func (o *StackTrieOptions) WithWriter(writer func(path []byte, hash common.Hash, blob []byte)) *StackTrieOptions {
+	o.Writer = writer
+	return o
+}
+
+// WithCleaner configures path cleaner within the options.
+func (o *StackTrieOptions) WithCleaner(cleaner func(path []byte)) *StackTrieOptions {
+	o.Cleaner = cleaner
+	return o
+}
+
+// SkipBoundary configures whether the left and right boundary nodes are filtered
+// for committing, along with a onBoundary callback invoked whenever the boundary
+// nodes are met.
+func (o *StackTrieOptions) SkipBoundary(skipLeft, skipRight bool, onBoundary func(path []byte, hash common.Hash, blob []byte)) *StackTrieOptions {
+	o.SkipLeftBoundary = skipLeft
+	o.SkipRightBoundary = skipRight
+	o.OnBoundary = onBoundary
+	return o
+}
 
 // StackTrie is a trie implementation that expects keys to be inserted
 // in order. Once it determines that a subtree will no longer be inserted
 // into, it will hash it and free up the memory it uses.
 type StackTrie struct {
-	writeFn NodeWriteFunc // function for committing nodes, can be nil
-	root    *stNode
-	h       *hasher
+	options      *StackTrieOptions
+	root         *stNode
+	h            *hasher
+	leftBoundary []byte
```

Comment:
If we store the leftBoundary in the struct, why not store the rigth boundary (or `last`) also? Basicaly, `first`, and `last` path that has been inserted. Then `hash` doesn't need to have it as an argument